### PR TITLE
FAI-456: CF: ExplainabilityService handling of CF requests

### PR DESCRIPTION
**Pull request by manstis**
See https://issues.redhat.com/browse/FAI-456

Many thanks for submitting your Pull Request :heart:! 

Please make sure that your PR meets the following requirements:

- [x] You have read the [contributors guide](https://github.com/kiegroup/kogito-runtimes#contributing-to-kogito)
- [x] Pull Request title is properly formatted: `KOGITO-XYZ Subject`
- [x] Pull Request title contains the target branch if not targeting master: `[0.9.x] KOGITO-XYZ Subject`
- [x] Pull Request contains link to the JIRA issue
- [x] Pull Request contains link to any dependent or related Pull Request
- [x] Pull Request contains description of the issue
- [x] Pull Request does not include fixes for issues other than the main ticket

<details>
<summary>
How to retest this PR or trigger a specific build:
</summary>

* <b>Pull Request</b>  
  Please add comment: <b>Jenkins retest this</b>
 
* <b>Quarkus LTS checks</b>  
  Please add comment: <b>Jenkins run LTS</b>

* <b>Native checks</b>  
  Please add comment: <b>Jenkins run native</b>

</details>

### CODE REVIEW

File: explainability/explainability-service/src/main/java/org/kie/kogito/explainability/CounterfactualExplainerProducer.java [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+
+import javax.enterprise.context.ApplicationScoped;
+import javax.enterprise.inject.Produces;
+
+import org.kie.kogito.explainability.local.counterfactual.CounterfactualExplainer;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+@ApplicationScoped
+public class CounterfactualExplainerProducer {
```
Comment (manstis) [inline explainability/explainability-service/src/main/java/org/kie/kogito/explainability/CounterfactualExplainerProducer.java]: @ruivieira This is a dummy implementation until your one comes along!

File: explainability/explainability-service/src/main/java/org/kie/kogito/explainability/handlers/CounterfactualExplainerProducer.java [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+
+import javax.enterprise.context.ApplicationScoped;
+import javax.enterprise.inject.Produces;
+
+import org.kie.kogito.explainability.local.counterfactual.CounterfactualExplainer;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+@ApplicationScoped
+public class CounterfactualExplainerProducer {
```
Comment (manstis) [inline explainability/explainability-service/src/main/java/org/kie/kogito/explainability/handlers/CounterfactualExplainerProducer.java:27]: @ruivieira This is my temporary producer for CF.

File: explainability/explainability-service/src/main/java/org/kie/kogito/explainability/handlers/CounterfactualExplainerServiceHandler.java [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+        this.explainer = explainer;
+    }
+
+    @Override
+    public boolean supports(BaseExplainabilityRequest request) {
+        return request instanceof CounterfactualExplainabilityRequest;
+    }
+
+    @Override
+    public BaseExplainabilityResultDto createSucceededResultDto(String executionId, CounterfactualResult result) {
```
Comment (manstis) [inline explainability/explainability-service/src/main/java/org/kie/kogito/explainability/handlers/CounterfactualExplainerServiceHandler.java:49]: @ruivieira You may want to change these to build the correct "success" object for CF.

File: explainability/explainability-core/src/main/java/org/kie/kogito/explainability/local/counterfactual/CounterfactualExplainer.java [hunk 4]
```diff
@@ -71,6 +72,10 @@
     public static final Consumer<CounterfactualSolution> defaultFinalConsumer =
             counterfactual -> logger.debug("Final counterfactual: {}", counterfactual.getEntities());
 
+    public CounterfactualExplainer() {
```
Comment (manstis) [inline explainability/explainability-core/src/main/java/org/kie/kogito/explainability/local/counterfactual/CounterfactualExplainer.java:75]: I believe zero-parameter constructors are required for CDI proxies.

File: explainability/explainability-service/src/main/java/org/kie/kogito/explainability/ExplanationServiceImpl.java [hunk 5]
```diff
@@ -45,47 +43,37 @@
...
-
     private static final Logger LOG = LoggerFactory.getLogger(ExplanationServiceImpl.class);
 
-    private final LocalExplainer<Map<String, Saliency>> localExplainer;
+    private final Instance<LocalExplainerServiceHandler<?>> localExplainers;
 
     @Inject
-    public ExplanationServiceImpl(LocalExplainer<Map<String, Saliency>> localExplainer) {
-        this.localExplainer = localExplainer;
+    public ExplanationServiceImpl(Instance<LocalExplainerServiceHandler<?>> localExplainers) {
```
Comment (manstis) [inline explainability/explainability-service/src/main/java/org/kie/kogito/explainability/ExplanationServiceImpl.java:51]: @ruivieira FYI.. here I inject all implementations of `LocalExplainerServiceHandler`.

`LocalExplainerServiceHandler` exists in the `-service` module and acts as a helper for `LocalExplainer` to primarily indicate which implementations support which explanations but also to provide factory methods to build the applicable "success" and "failure" messages. I did not add the methods to `LocalExplainer` itself as it meant `-core` needed to have additional dependencies on other modules or I move a bunch of things into `-core`. This decoupling seemed the cleanest approach.

Since the implementation of the various

File: explainability/explainability-service/src/main/java/org/kie/kogito/explainability/ExplanationServiceImpl.java [hunk 6]
```diff
@@ -45,47 +43,37 @@
...
             PredictionProvider predictionProvider) {
         LOG.debug("Explainability request with executionId {} for model {}:{}",
                 request.getExecutionId(),
                 request.getModelIdentifier().getResourceType(),
                 request.getModelIdentifier().getResourceId());
         try {
-            return localExplainer.explainAsync(getPrediction(request.getInputs(), request.getOutputs()), predictionProvider)
-                    .thenApply(input -> createSucceededResultDto(request.getExecutionId(), input))
-                    .exceptionally(e -> createFailedResultDto(request.getExecutionId(), e));
+            LocalExplainerServiceHandler<?> localExplainer =
```
Comment (manstis) [inline explainability/explainability-service/src/main/java/org/kie/kogito/explainability/ExplanationServiceImpl.java:64]: @ruivieira This is in essence the magic.

A `LocalExplainerServiceHandler` is looked up for the explanation request and invoked accordingly.

File: explainability/explainability-service/src/main/java/org/kie/kogito/explainability/ExplanationServiceImpl.java [hunk 7]
```diff
@@ -45,47 +43,37 @@
...
+                    getLocalExplainer(request).orElseThrow(() -> new IllegalArgumentException("LocalExplainerServiceHandler could not be found for '" + request.getClass().getName() + "'"));
+            return localExplainer.explainAsyncWithResults(getPrediction(request.getInputs(),
+                    request.getOutputs()),
+                    predictionProvider,
+                    request.getExecutionId());
         } catch (Exception e) {
-            return CompletableFuture.completedFuture(createFailedResultDto(request.getExecutionId(), e));
+            return CompletableFuture.completedFuture(ExceptionResultDto.buildFailed(request.getExecutionId(), LocalExplainerServiceHandler.FAILED_STATUS_DETAILS));
```
Comment (manstis) [inline explainability/explainability-service/src/main/java/org/kie/kogito/explainability/ExplanationServiceImpl.java:71]: Should an exception occur a `LocalExplainerServiceHandler` may not have been found so I added a new generic _result_ `ExceptionResultDto`. I guess I could improve this to first check if `localExplainer` was in use and, if so, use that to build the failure message. WDYT?

File: explainability/explainability-service/src/main/java/org/kie/kogito/explainability/handlers/LimeExplainerServiceHandler.java [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+        this.explainer = explainer;
+    }
+
+    @Override
+    public boolean supports(BaseExplainabilityRequest request) {
+        return request instanceof LIMEExplainabilityRequest;
+    }
+
+    @Override
+    public BaseExplainabilityResultDto createSucceededResultDto(String executionId, Map<String, Saliency> result) {
```
Comment (manstis) [inline explainability/explainability-service/src/main/java/org/kie/kogito/explainability/handlers/LimeExplainerServiceHandler.java:52]: These methods were simply copied from the old `ExplainationServiceImpl`.

**Comment by r00ta** [comment]
Hey @manstis , I'll have a look once https://github.com/kiegroup/kogito-apps/pull/747 is merged so that I can have a proper diff display :+1:

**Comment by manstis** [comment]
Jenkins retest this

**Comment by manstis** [comment]
@ruivieira @r00ta @jiripetrlik This PR is ready for review.

It covers "up to" entry into the `CounterfactualExplainer` but no response at the moment.

@ruivieira is working on another PR to refactor the explainer and we are meeting somewhere in the middle.

Once this PR and Rui's are merged I can work on completing the return (solutions) trip.

**Comment by manstis** [comment]
Jenkins retest this

**Comment by manstis** [comment]
Jenkins retest this

**Comment by manstis** [comment]
Hi @r00ta @ruivieira @jiripetrlik it would be nice if you could find a little time to review this please.

**Comment by manstis** [comment]
Rebased with `master`.

**Comment by r00ta** [comment]
> Hi @r00ta @ruivieira @jiripetrlik it would be nice if you could find a little time to review this please.

Hi @manstis , I apologize, i did not have time to look into this. I will do that today or tomorrow

File: explainability/explainability-api/src/main/java/org/kie/kogito/explainability/api/CounterfactualExplainabilityResultDto.java [hunk 9]
```diff
@@ -15,23 +15,97 @@
...
+    private String counterfactualId;
+
+    @JsonProperty(IS_VALID_FIELD)
+    private Boolean isValid;
+
+    @JsonProperty(INPUTS_FIELD)
+    private Map<String, TypedValue> inputs;
+
+    @JsonProperty(OUTPUTS_FIELD)
+    private Map<String, TypedValue> outputs;
```
Comment (r00ta) [inline explainability/explainability-api/src/main/java/org/kie/kogito/explainability/api/CounterfactualExplainabilityResultDto.java:50]: Given that this is the result of the Counterfactual analysis, what are we going to put inside `inputs` and `outputs`?

File: explainability/explainability-api/src/main/java/org/kie/kogito/explainability/api/LIMEExplainabilityResultDto.java [hunk 10]
```diff
@@ -21,7 +21,7 @@
 
 public class LIMEExplainabilityResultDto extends BaseExplainabilityResultDto {
 
-    public static final String EXPLAINABILITY_TYPE_NAME = "LIME";
+    public static final String EXPLAINABILITY_TYPE_NAME = "lime";
```
Comment (r00ta) [inline explainability/explainability-api/src/main/java/org/kie/kogito/explainability/api/LIMEExplainabilityResultDto.java:24]: just for my curiosity, why this change?

File: explainability/explainability-core/src/main/java/org/kie/kogito/explainability/local/counterfactual/CounterfactualExplainer.java [hunk 11]
```diff
@@ -71,6 +72,10 @@
...
+        this(null, null, Collections.emptyList(), Collections.emptyList(), null, null, null, null);
+    }
+
```
Comment (r00ta) [inline explainability/explainability-core/src/main/java/org/kie/kogito/explainability/local/counterfactual/CounterfactualExplainer.java]: Do we need this?

File: explainability/explainability-api/src/main/java/org/kie/kogito/explainability/api/CounterfactualExplainabilityRequestDto.java [hunk 12]
```diff
@@ -27,8 +29,14 @@
...
 
+    public static final String COUNTERFACTUAL_ID_FIELD = "counterfactualId";
+
     public static final String SEARCH_DOMAINS_FIELD = "searchDomains";
 
+    @JsonProperty(COUNTERFACTUAL_ID_FIELD)
+    @NotNull(message = "counterfactualId must be provided.")
+    private String counterfactualId;
+
     @JsonProperty(SEARCH_DOMAINS_FIELD)
```
Comment (r00ta) [inline explainability/explainability-api/src/main/java/org/kie/kogito/explainability/api/CounterfactualExplainabilityRequestDto.java:40]: I might be wrong but why do we put the `CFID` + `ExecutionId` in the `CounterfactualExplainabilityRequestDto`? iiuc the `CounterfactualExplainer` of @ruivieira ignores it and sets it when a solution/result is found (the file with this comment https://github.com/kiegroup/kogito-apps/pull/780/files#r622785167 - don't look at the comment itself but at the file it references to)

File: explainability/explainability-core/src/main/java/org/kie/kogito/explainability/model/Prediction.java [hunk 13]
```diff
@@ -24,10 +26,19 @@
...
+    private final PredictionFeatureDomain domain;
+
     public Prediction(PredictionInput input,
             PredictionOutput output) {
+        this(input, output, new PredictionFeatureDomain(Collections.emptyList()));
+    }
+
+    public Prediction(PredictionInput input,
+            PredictionOutput output,
+            PredictionFeatureDomain domain) {
```
Comment (r00ta) [inline explainability/explainability-core/src/main/java/org/kie/kogito/explainability/model/Prediction.java]: See the comment here https://github.com/kiegroup/kogito-apps/pull/780/files#r622785167 , I'm a little bit concerned that we abuse the use of `output` to store the "goals" of the CF request

**Review by ruivieira — APPROVED** [review]
@manstis lgtm, thanks!

Not a comment, just a note to all, that some `org.kie.kogito.explainability.model.*` (namely `Prediction`) classes will probably change with FAI-467.

File: trusty/trusty-service/trusty-service-common/src/main/java/org/kie/kogito/trusty/service/common/handlers/CounterfactualExplainerServiceHandler.java [hunk 14]
```diff
@@ -0,0 +1,82 @@
...
+    }
+
+    @Override
+    public CounterfactualExplainabilityResult explainabilityResultFrom(CounterfactualExplainabilityResultDto dto, Decision decision) {
+        return new CounterfactualExplainabilityResult(dto.getExecutionId(),
+                dto.getCounterfactualId(),
+                "solutionId",
+                statusFrom(dto.getStatus()),
+                dto.getStatusDetails());
+    }
```
Comment (r00ta) [inline trusty/trusty-service/trusty-service-common/src/main/java/org/kie/kogito/trusty/service/common/handlers/CounterfactualExplainerServiceHandler.java:62]: why `solutionId` is the `statusDetails` of the instance `CounterfactualExplainabilityResult`?

**Review by r00ta — COMMENTED** [review]
Hi @manstis , thank you for the PR. Let me know what do you think about the comments. 
I did not review the tests, I leave them to @jiripetrlik

File: explainability/explainability-api/src/main/java/org/kie/kogito/explainability/api/LIMEExplainabilityResultDto.java (hunk 10 above)
Comment (manstis) [reply explainability/explainability-api/src/main/java/org/kie/kogito/explainability/api/LIMEExplainabilityResultDto.java:24]: No good reason, other than lowercase is [consistent](https://github.com/kiegroup/kogito-apps/pull/752/files#diff-405bc051141284f4df4d316e0300b957cca90e841c7ac79690156ebfb13d811eR29) with `counterfactual`.

File: explainability/explainability-core/src/main/java/org/kie/kogito/explainability/local/counterfactual/CounterfactualExplainer.java (hunk 11 above)
Comment (manstis) [reply explainability/explainability-core/src/main/java/org/kie/kogito/explainability/local/counterfactual/CounterfactualExplainer.java]: I believe a zero-parameter constructor is normally required by CDI.....

.... but since we have a producer for the explainer it probably is not necessary. Let me see if I can remove it.

File: explainability/explainability-api/src/main/java/org/kie/kogito/explainability/api/CounterfactualExplainabilityRequestDto.java (hunk 12 above)
Comment (manstis) [reply explainability/explainability-api/src/main/java/org/kie/kogito/explainability/api/CounterfactualExplainabilityRequestDto.java:40]: The sequence is this: 

UI -> [request CF] -> TrustyService -> [create `counterfactualId`, store requirement, send CE] -> ExplainabilityService -> [create solutions, intermediate and final, send them back] -> TrustyService -> [store results using `executionId`/`counterfactualId` and `solutionId` for each result] -> UI query for results using `executionId`/`counterfactualId`.

So, the request contains the `counterfactualId` (that was created by `TrustyService`) and it is included in the results. `TrustyService` can then correlate the results with the request. I think this is different to the `counterfactualId` @ruivieira has been using which, from my perspective is more a `solutionId` (that you have commented on later too).

Note, @ruivieira code sets an _ID_ on a class called `CounterfactualSolution` representing a single CF solution. 

This is where I [started](https://github.com/kiegroup/kogito-apps/pull/752/files#diff-5a5cba9eac086624633ad40846cea6a2019850b3b1a8ba8ba8ddc64a6fa92290R130) the return trip (but it is incomplete in this PR and has no support for intermediate results.. that is also part of some other work @ruivieira is doing too!)

File: explainability/explainability-api/src/main/java/org/kie/kogito/explainability/api/CounterfactualExplainabilityResultDto.java (hunk 9 above)
Comment (manstis) [reply explainability/explainability-api/src/main/java/org/kie/kogito/explainability/api/CounterfactualExplainabilityResultDto.java:50]: The return trip is work in progress on my laptop. What is in this PR is "just enough" to connect some of the dots in the whole round trip. The `inputs` and `outputs` are the values CF returned as a possible solution. i.e. although the User requested output X is, for example 100, there may only be a CF solution where output X is 101. I suggest you do not focus on reviewing the "return trip" for this PR.

File: trusty/trusty-service/trusty-service-common/src/main/java/org/kie/kogito/trusty/service/common/handlers/CounterfactualExplainerServiceHandler.java (hunk 14 above)
Comment (manstis) [reply trusty/trusty-service/trusty-service-common/src/main/java/org/kie/kogito/trusty/service/common/handlers/CounterfactualExplainerServiceHandler.java:62]: Yes, that is wrong.. my work locally is for the "return trip" and `CounterfactualExplainabilityResult` contains `executionId`, `counterfactualId`, `solutionId`, `inputs` and `outputs`. This PR only contains minimal return trip stuff some of which is place holders (i.e.  `CounterfactualExplainerServiceHandler` needed _an_ implementation of this method to compile).

File: explainability/explainability-core/src/main/java/org/kie/kogito/explainability/local/counterfactual/CounterfactualExplainer.java (hunk 11 above)
Comment (manstis) [reply explainability/explainability-core/src/main/java/org/kie/kogito/explainability/local/counterfactual/CounterfactualExplainer.java]: I will look when I rebase this PR with @ruivieira stateless-explainer commits after it has been merged.

File: explainability/explainability-core/src/main/java/org/kie/kogito/explainability/model/Prediction.java (hunk 13 above)
Comment (manstis) [reply explainability/explainability-core/src/main/java/org/kie/kogito/explainability/model/Prediction.java]: See https://issues.redhat.com/browse/FAI-480

File: explainability/explainability-api/src/main/java/org/kie/kogito/explainability/api/CounterfactualExplainabilityResultDto.java (hunk 9 above)
Comment (manstis) [reply explainability/explainability-api/src/main/java/org/kie/kogito/explainability/api/CounterfactualExplainabilityResultDto.java:50]: See https://issues.redhat.com/browse/FAI-480

File: explainability/explainability-core/src/main/java/org/kie/kogito/explainability/local/counterfactual/CounterfactualExplainer.java (hunk 11 above)
Comment (manstis) [reply explainability/explainability-core/src/main/java/org/kie/kogito/explainability/local/counterfactual/CounterfactualExplainer.java]: It was not needed. Removed.

File: explainability/explainability-api/src/main/java/org/kie/kogito/explainability/api/CounterfactualExplainabilityRequestDto.java (hunk 12 above)
Comment (r00ta) [reply explainability/explainability-api/src/main/java/org/kie/kogito/explainability/api/CounterfactualExplainabilityRequestDto.java:40]: Hi @manstis , thank you for the detailed clarification: now I understand. As we already discussed on the chat the `counterfactualId` in this PR https://github.com/kiegroup/kogito-apps/pull/780 has been renamed to `solutionId`. All good :)

**Review by r00ta — APPROVED** [review]
Thank you @manstis for the nice job and for creating the tickets about my comments. If the tests are ok for @jiripetrlik I think we can move forward with this PR, work on the tickets and get closer to the final integration :)

**Comment by manstis** [comment]
@jiripetrlik Could you please review.

**Comment by manstis** [comment]
@danielezonca This will need your review too. Thank-you.

File: explainability/explainability-service/src/test/java/org/kie/kogito/explainability/ConversionUtilsTest.java [hunk 15]
```diff
@@ -243,4 +252,70 @@ private void commonTypeValuePairJson(JsonNode input, Type type, Object value) {
...
+        CategoricalFeatureDomain categoricalFeatureDomain = (CategoricalFeatureDomain) featureDomain;
+        assertEquals(2, categoricalFeatureDomain.getCategories().size());
+        assertTrue(categoricalFeatureDomain.getCategories().containsAll(List.of("White", "Black")));
+        assertNull(categoricalFeatureDomain.getLowerBound());
+        assertNull(categoricalFeatureDomain.getUpperBound());
+    }
+
+    @Test
+    void testFromFeatureList() {
+
```
Comment (jiripetrlik) [inline explainability/explainability-service/src/test/java/org/kie/kogito/explainability/ConversionUtilsTest.java]: Is this intentional? What is the purpose of this test?

File: trusty/trusty-service/trusty-service-common/src/main/java/org/kie/kogito/trusty/service/common/handlers/BaseExplainerServiceHandler.java [hunk 16]
```diff
@@ -0,0 +1,45 @@
...
+    }
+
+    protected ExplainabilityStatus statusFrom(org.kie.kogito.explainability.api.ExplainabilityStatus status) {
+        switch (status) {
+            case SUCCEEDED:
+                return ExplainabilityStatus.SUCCEEDED;
+            case FAILED:
+                return ExplainabilityStatus.FAILED;
+        }
+        return null;
```
Comment (jiripetrlik) [inline trusty/trusty-service/trusty-service-common/src/main/java/org/kie/kogito/trusty/service/common/handlers/BaseExplainerServiceHandler.java]: Can we log some warning if there is some unknown state?

File: explainability/explainability-service/src/test/java/org/kie/kogito/explainability/ConversionUtilsTest.java (hunk 15 above)
Comment (manstis) [reply explainability/explainability-service/src/test/java/org/kie/kogito/explainability/ConversionUtilsTest.java]: Good catch.. removed (it will be added back, with implementation, in a subsequent PR).

File: trusty/trusty-service/trusty-service-common/src/main/java/org/kie/kogito/trusty/service/common/handlers/BaseExplainerServiceHandler.java (hunk 16 above)
Comment (manstis) [reply trusty/trusty-service/trusty-service-common/src/main/java/org/kie/kogito/trusty/service/common/handlers/BaseExplainerServiceHandler.java]: There are only two possible values in both enumeration.

I have refactored this to an `if.. else` and avoided the possible `null` return value.

**Comment by manstis** [comment]
@jiripetrlik Thank-you for the review. I have pushed a new commit with the changes.

**Review by danielezonca — APPROVED** [review]
LGTM 👍

**Review by jiripetrlik — APPROVED** [review]
@manstis
Thank you for your changes. Looks good to me now.
